### PR TITLE
BP003: Task Life-Cycle blueprint [v4]

### PR DESCRIPTION
**Pull request by clebergnu**
This introduces the high level architecture of the N(ext) Runner
integration with an Avocado Job, based on Tasks' Life-Cycle.

Signed-off-by: Willian Rampazzo <willianr@redhat.com>
Signed-off-by: Cleber Rosa <crosa@redhat.com>

---

Changes from v3 (#4075):
* Added reviewers
* Changed status to WIP
* Added issues to "Related Issues" section
* Moved example task life cycle implementation to `examples/nrunner/task-state-machine.py`
* Renamed TaskLifeCycle to TaskStateMachine
* Added `TASK_` prefix where it was missing
* Various typos and wording improvements

Changes from v2 (#4061):
* New section "Goals of this BluePrint", with some information that was previously under "Motivation"
* Removal of the tasks_preparing queue
* A task being triaged, will now only check for the environment ability to run it, it will no longer consider the fulfillment of requirements for the LTS version.
* Merged all final phases (FAILED and FINISHED) into FINISHED.
* Added introductory text to "Requirements" section.
* Added an implementation example
* Simplified the iterations documentation example, replacing it with a more real implementation example
* Various typos and wording improvements

Changes from v1 (#4016):
* Focused on clearer and better defined Motivation and Requirements section
* Combined superfluous task states

### CODE REVIEW

File: docs/source/blueprints/BP003.rst [hunk 1]
```diff
@@ -0,0 +1,573 @@
...
+Users should not be required to learn about the N(ext) Runner features
+to use it just as an alternative to the current runner implementation.
+
+Related Issues
+**************
+
+Current issues that are expected to be solved when this blueprint is
+implemented:
+
+#. Have a passive Task Status collection server implementation.
```
Comment (beraldoleal) [inline docs/source/blueprints/BP003.rst:523]: Just to be picky: A "passive task status collection server implementation" is not an issue, right?

File: docs/source/blueprints/BP003.rst [hunk 2]
```diff
@@ -0,0 +1,573 @@
...
+
+#. Have a Task Life Cycle / State Machine implementation.
```
Comment (beraldoleal) [inline docs/source/blueprints/BP003.rst:525]: Here too.

File: docs/source/blueprints/BP003.rst [hunk 3]
```diff
@@ -0,0 +1,573 @@
...
+   This is inteded to be used in place or in addition of the status
+   messages from Tasks, when they failed to be generated by tasks
+   or received by the Task Status collection server.
+
+#. Have Spawner features to destroy (best effort) stray Tasks.
+
+#. Fully integrate the N(ext) Runner integrate into the Avocado Job
+   and command line app, that is, as a general rule all features of
+   the current runner should be present when the N(ext) Runner is
+   used in a job.
```
Comment (beraldoleal) [inline docs/source/blueprints/BP003.rst:537]: Yep, those are not "issues", those are the 'outcome', the benefits, IIUC.

File: docs/source/blueprints/BP003.rst (hunk 1 above)
Comment (clebergnu) [reply docs/source/blueprints/BP003.rst:523]: Well, I was trying to follow the model on BP001, such as:

```
#. Create a simple but effective type system for configuration files and
   argument options.
```

That is, with issues that will be created (in GH) and solved when this BP is considered implemented.

Comment (beraldoleal) [reply docs/source/blueprints/BP003.rst:523]: Ohh I see. Ok, makes sense. sorry about the noise.

Comment (willianrampazzo) [reply docs/source/blueprints/BP003.rst:523]: Maybe this become more clear if it is written just like the issue: "Implement the passive Task Status collection server."

File: docs/source/blueprints/BP003.rst (hunk 2 above)
Comment (willianrampazzo) [reply docs/source/blueprints/BP003.rst:525]: "Implement the Task Life Cycle / State Machine."

File: docs/source/blueprints/BP003.rst [hunk 4]
```diff
@@ -0,0 +1,573 @@
...
+#. Have a Task Life Cycle / State Machine implementation.
+
+#. Have Spawner features to check the status (alive or not) for Tasks.
+   This is inteded to be used in place or in addition of the status
+   messages from Tasks, when they failed to be generated by tasks
+   or received by the Task Status collection server.
+
+#. Have Spawner features to destroy (best effort) stray Tasks.
+
+#. Fully integrate the N(ext) Runner integrate into the Avocado Job
```
Comment (willianrampazzo) [inline docs/source/blueprints/BP003.rst:534]: Double `integrate` in this line.

**Comment by willianrampazzo** [comment]
> Except for @beraldoleal's comments about the `Issues` section, nothing else from my side.

Feel free to force push the corrections, if you make some, as they are minor text changes, and let me know.

**Comment by clebergnu** [comment]
> > Except for @beraldoleal's comments about the `Issues` section, nothing else from my side.
> 
> Feel free to force push the corrections, if you make some, as they are minor text changes, and let me know.

Done as 11024c70cee58e34baea72011a95ada4cd5dbe3b.

**Pull request merged**